### PR TITLE
Explosions, TNT, Flint&Steel implementation

### PR DESCRIPTION
**Pull request by Tonodus**
This is finished now, only missing things are:
- [ ] check whether the block is flammable (with Flint&Steel / by TNT)
- [ ] use the block's blast resistance
  => probably finished by #541 

This requires values I'd prefer to be stored along with block's hardness, light level, and so on, so I'm waiting for the implementation of these to finish the things mentioned above.

Also please notice the survival bug I mentioned: #330 
There's another "bug" in GlowPlayer: A player's velocity won't be set to zero sometime, so if you add a velocity to a player it will stay there forever. Since an explosion "throws players away" by adding values to the player's velocity, the velocity will be higher and higher by every explosion.
That means the first explosion creates the right values for velocity, but after the second explosion the velocity will be double as high as normal, since the player keeps the velocity from the first explosion.
This won't be fixed in this PR, as it's a general bug in GlowEntity.

What's done:
- [x] add ItemType ItemFlintAndSteel
- [x] create explosions and drop the related blocks
- [x] setting blocks of explosions on fire
- [x] damage nearby entities [DONE, needs some implementation in GlowWorld]
- [x] set new velocity for affected entities
- [x] detect and fire near TNT
- [x] send explosion sound
- [x] send explosion particles
- [x] define and send explosion-packet (ExplosionMessage.java/ExplosionCodec.java)

---

_Edited by turt2live_

**Related Links:**
- Ticket: #203 (damage)
- Ticket: #139 (entities)
- No ticket for behaviour (none needed)

### CODE REVIEW

**Comment by ZephireNZ** [comment]
Woah, awesome work! If anyone is wondering, here is an explaination for algorithm Vanilla used ( that seems to be replicated in this PR): https://www.youtube.com/watch?v=Tos9cSLJNTg

**Comment by Tonodus** [comment]
Good video, a text based explaination is here: http://minecraft.gamepedia.com/Explosion

File: src/main/java/net/glowstone/Explosion.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+            GlowBlock block = (GlowBlock) world.getBlockAt(current);
+
+            if (block.getType() != Material.AIR) {
+                double sub = getBlastDurability(block);
+                sub /= 5;
+                sub += 0.3F;
+                sub *= 0.3F;
+                currentPower -= sub;
+
+                result.add(new WorldlessLocation(block.getX(), block.getY(), block.getZ()));
```
Comment (Tonodus) [inline src/main/java/net/glowstone/Explosion.java]: TODO: Need to check, whether the block breaks!

File: src/main/java/net/glowstone/Explosion.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            //send large explosion
+        }
+    }
+
+    private void playOutExplodeSmoke(Location location) {
+
+    }
+
+
+    private static final class WorldlessLocation {
```
Comment (SpaceManiac) [inline src/main/java/net/glowstone/Explosion.java]: Vector and BlockVector already exist for this purpose.

Comment (Tonodus) [inline src/main/java/net/glowstone/Explosion.java]: @SpaceManiac: You're right, changed it.

**Comment by Aaron1011** [comment]
@Tonodus: I think you meant 'blast' instead of 'blaze'

**Comment by Tonodus** [comment]
@Aaron1011: You're right. I only misspelled this in the PR's description, didn't I?

**Comment by Aaron1011** [comment]
@Tonodus: Yeah, that was the only place.

File: src/main/java/net/glowstone/GlowWorld.java [hunk 3]
```diff
@@ -416,6 +416,21 @@ public void pulse() {
...
+     * @return a value between 0 and 1, where 0 = all rays blocked and 1 = all rays unblocked
+     */
+    public float rayTrace(Location location, GlowEntity entity) {
+        /**
+         * To calculate this step through the entity's bounding box and check whether the ray to the point in the bounding box is blocked.
+         * Return (unblockedRays / allRays)
+         */
+        return 0;
+    }
+
```
Comment (Wild1145) [inline src/main/java/net/glowstone/GlowWorld.java:460]: Probably a bit much white space here ;)

**Comment by Wild1145** [comment]
Just double check on the formatting, some files look to have a bit too much white space, though I'd need to check the code formatting for this project to be sure.

**Comment by turt2live** [comment]
@Tonodus This is looking pretty inactive code-wise. If after the merge conflicts are resolved this PR is done, please remove "WIP" from the title to have it reviewed. 

If this is still a work in progress, please comment back stating so. 

Thanks!

(PS: Remember to limit your scope. Don't need to implement everything fire-related right now)

**Comment by dequis** [comment]
PR body says:

> This is finished now, only missing things are [...]
> I'm waiting for the implementation of these to finish the things mentioned above.

I'd say go ahead and test this. Sounds fun

**Comment by Tonodus** [comment]
I removed [WIP] since this is functional and can be tested.
However, please consider the bug mentioned in the PR description and that all blocks have a blast resistance from dirt (2.5).

> Don't need to implement everything fire-related right now

If this is ok, I would leave the fire implementation as it is in this PR and don't add the possibility to set sides of block on fire. This is anyway not a full fire implementation (since no checks for burnable blocks, blocks don't ignite nearby blocks), only Flint&Steel implementation for TNT.

File: src/main/java/net/glowstone/Explosion.java [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.block.BlockIgniteEvent;
+import org.bukkit.event.entity.EntityExplodeEvent;
+import org.bukkit.util.BlockVector;
+import org.bukkit.util.Vector;
+
+import java.util.*;
+
+public class Explosion {
+    public static final int POWER_TNT = 4;
```
Comment (turt2live) [inline src/main/java/net/glowstone/Explosion.java:26]: This screams enum

File: src/main/java/net/glowstone/GlowWorld.java [hunk 5]
```diff
@@ -422,7 +427,24 @@ public void pulse() {
     }
 
     /**
+     * Calculates how much the rays from the location to the entity's bounding box is blocked.
+     *
+     * @param location The location for the rays to start
+     * @param entity The entity that's bounding box is the ray's end point
+     * @return a value between 0 and 1, where 0 = all rays blocked and 1 = all rays unblocked
+     */
+    public float rayTrace(Location location, GlowEntity entity) {
```
Comment (turt2live) [inline src/main/java/net/glowstone/GlowWorld.java:451]: This looks incomplete... what is it supposed to do and is it needed?

File: src/main/java/net/glowstone/block/blocktype/BlockTNT.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package net.glowstone.block.blocktype;
+
+import net.glowstone.block.GlowBlock;
+import org.bukkit.Material;
+import org.bukkit.Sound;
+import org.bukkit.World;
+import org.bukkit.entity.EntityType;
+
+public class BlockTNT extends BlockType {
+    public void explodeTNTBlock(GlowBlock tntBlock) {
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockTNT.java]: I would personally shorten this to `explodeBlock`

File: src/main/java/net/glowstone/block/itemtype/ItemFlintAndSteel.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                fireNetherPortal();
+                return true;
+            //TODO check for non flammable blocks
+            default:
+                return setBlockOnFire(player, target, face, holding, clickedLoc);
+        }
+    }
+
+    private void fireNetherPortal() {
+        //TODO
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/itemtype/ItemFlintAndSteel.java]: Maybe add a message to this for future contributors scanning the list of `TODO` instances in their IDE

File: src/main/java/net/glowstone/block/itemtype/ItemTool.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Override
+    public final void rightClickBlock(GlowPlayer player, GlowBlock target, BlockFace face, ItemStack holding, Vector clickedLoc) {
+        if (onToolRightClick(player, holding, target, face, clickedLoc)) {
+            damageTool(player, holding);
+        }
+    }
+
+    private void damageTool(GlowPlayer player, ItemStack holding) {
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/itemtype/ItemTool.java:27]: Note: This might need to be extended later for other components down the road. It may also need to be refactored in general down the road. (no action to be done in this PR, just a note for the future)

File: src/main/java/net/glowstone/entity/GlowExplosive.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        this.yield = yield;
+    }
+
+    @Override
+    public float getYield() {
+        return yield;
+    }
+
+    @Override
+    public void setIsIncendiary(boolean isIncendiary) {
```
Comment (turt2live) [inline src/main/java/net/glowstone/entity/GlowExplosive.java]: Remove the `is` (twice)

File: src/main/java/net/glowstone/entity/GlowExplosive.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
+package net.glowstone.entity;
+
+import org.bukkit.Location;
+import org.bukkit.entity.Explosive;
+
+public abstract class GlowExplosive extends GlowEntity implements Explosive {
+    private float yield;
+    private boolean isIncendiary;
```
Comment (turt2live) [inline src/main/java/net/glowstone/entity/GlowExplosive.java]: Adding `is` is not really needed.

File: src/main/java/net/glowstone/Explosion.java [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+            current.add(direction);
+            currentPower -= 0.225f;
+        }
+    }
+
+    private void handleBlockExplosion(GlowBlock block) {
+        if (block.getType() == Material.AIR) {
+            return;
+        } else if (block.getType() == Material.TNT) {
+            ((BlockTNT) itemTable.getBlock(Material.TNT)).explodeTNTBlock(block);
```
Comment (turt2live) [inline src/main/java/net/glowstone/Explosion.java]: From the wiki: `Its fuse lasts 40 redstone ticks (4 seconds/80 game ticks) if activated by redstone or fire, or a random number between 10 to 30 game ticks (.5 to 1.5 seconds) if it's destroyed by an explosion.` 

In other words, right here the number of ticks needs to somehow be passed down to the explosion objects (`setFuseTicks`). My recommendation is to add an overloaded method to `explodeBlock` that accepts a boolean: `blownUp` (name is just a suggestion). If that boolean is true, this 'random' tick count is applied, otherwise the standard is used. I would almost go as far to say that the boolean should further propagate to the explosion object that holds the fuse ticks so that magic numbers aren't all over the place.

**Comment by turt2live** [comment]
@Tonodus 

> If this is ok, I would leave the fire implementation as it is in this PR and don't add the possibility to set sides of block on fire. This is anyway not a full fire implementation (since no checks for burnable blocks, blocks don't ignite nearby blocks), only Flint&Steel implementation for TNT.

Having this PR only contain a way to set TNT off is perfectly fine as it helps test the TNT. Actual implementation of "can you burn that" and such is out of scope of this PR. 

As for block hardness and all that, I encourage you to open a second PR for that as soon as this is pulled :)

File: src/main/java/net/glowstone/Explosion.java (hunk 4 above)
Comment (Tonodus) [inline src/main/java/net/glowstone/Explosion.java:26]: This can't be an enum, since the bukkit API allows plugins to set an individual power, not only one of this constants.

File: src/main/java/net/glowstone/entity/GlowExplosive.java (hunk 9 above)
Comment (Tonodus) [inline src/main/java/net/glowstone/entity/GlowExplosive.java]: The method name is defined in the bukkit api, so it shouldn't be changed.

File: src/main/java/net/glowstone/GlowWorld.java (hunk 5 above)
Comment (Tonodus) [inline src/main/java/net/glowstone/GlowWorld.java:451]: MC contains several rayTracing methods. All this methods check, whether 'ray(s)' going from a given position in a given direction are blocked by blocks (and maybe entities).
In this case the methods checks how much the entity is affected by the explosion: If you can 'see' the entity from the explosion centre, apply full damage. If you can only see the entity partly, apply only partly damage.

Comment (turt2live) [inline src/main/java/net/glowstone/GlowWorld.java:451]: Then this should have either a `TODO` stating that when N is implemented, this is to be updated or a comment explaining why this is a constant if this is not waiting on anything else to be implemented.

**Comment by Tonodus** [comment]
I added the things you mentioned:
- Added TODO
- TNT will explode after a random number between 10 and 30 ticks now if activated by Explosion
- removed the `is`
- method is now `explodeBlock`

File: src/main/java/net/glowstone/Explosion.java [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+
+    private float calculateStartPower() {
+        float rand = random.nextFloat();
+        rand *= 0.6F; //(max - 0.7)
+        rand += 0.7; //min
+        return rand * power;
+    }
+
+    private double getBlastDurability(GlowBlock block) {
+        return 2.5; //Dirt
```
Comment (turt2live) [inline src/main/java/net/glowstone/Explosion.java]: Needs a TODO or ghost implementation

**Comment by dequis** [comment]
Simple conflicts in ItemTable:
- `import *` - keep from master
- Added items - keep both sides

Merged branch here: https://github.com/dequis/Glowstone/compare/pr-284

**Comment by turt2live** [comment]
@Tonodus Please don't hurt the creative players and then this is ready :)

**Comment by PaulBGD** [comment]
@turt2live actually, this is using a custom damage thing. It should be calling ent.damage(damage, cause). That method should be handling damage.

**Comment by turt2live** [comment]
@Tonodus The physics are a bit off, but that can be fixed in a later PR.

However, blowing up a tree causes the leaves to disappear, but not really. The client seems to get more information than it should about blocks that went missing. I tested this further by creating a giant cube of leaves and blowing up a single part of it. When placing blocks back on top of the missing blocks some returned to leaves, indicating that the server and client were mismatched.

**Comment by Tonodus** [comment]
Rebased against master and (if you haven't already noticed) fixed the bug you mentioned above.

**Comment by turt2live** [comment]
@Tonodus Noted some issues with the PR:
- Strange issues occur when placing and igniting TNT in survival mode
  - Carries over to creative mode when switching survival -> creative
  - Reproducible case: Place TNT in survival, ignite in survival.

I additionally fixed the spawn position myself. There has been a PR made to your branch with my fix. It also includes fixes for some javadocs and an added constructor to Explosion for ease of use in the future. https://github.com/Tonodus/Glowstone/pull/1

**Comment by Tonodus** [comment]
@turt2live 

Im conviced the issue you  mentioned is caused by #330 and not in the scope of the pr.
The TNT disappears only client-side btw.

**Comment by turt2live** [comment]
I would agree, thanks for the information.

**Comment by Tonodus** [comment]
Rebased on top of master.

**Comment by SpaceManiac** [comment]
Merged in 6c6b63e69d3aca711f727e37a2f0f0cd43b5647f.
